### ISSUE TITLE
reorganize JS code

### ISSUE THREAD

**Issue by aliceriot** [mitodl/micromasters#920]
@gsidebo has suggested many time that `util` is not a great name for a directory that contains a bunch of somewhat unrelated functions and react components. I think we should think a little bit about how best to organize things. Currently the `static/js` directory looks like this:

```
├── actions
├── components
│   ├── dashboard
│   ├── inputs
│   ├── search
│   └── utils
├── containers
├── flow
├── public
│   └── components
├── reducers
├── store
└── util
```

I think we should have some discussion about how best to reorganize it, and what would make sense. As a way to get the discussion going, here's one proposal:

```
├── actions
├── components
│   ├── dashboard
│   ├── inputs
│   ├── search
│   └── utils -> shared
├── containers
├── flow
├── public
│   └── components
├── reducers
├── store
└── util -> lib
```

So basically renaming the `utils` directory to `lib`, and renaming `components/utils` to `components/shared`. We could also move the few component-ey things that are currently in `util` into `components/shared`, and maybe think about gradually refactoring some more of the methods that return components which live in `util` into full fledged components (not necessarily classes, but living in their own file, having a name, having separate tests, etc).

Then `lib` would really be a bit more focused and would only really contain libraries we've written, such as the validation library, the api library, sorting functions, etc.

What do you think @noisecapella @gsidebo ?

**Comment by gsidebo** [comment]
thanks for getting this going! just want to clarify that i'm not against a directory or file named 'util' as long as it contains utility functions 😄 . take static/js/util/util.js for example. `asPercent`, `callFunctionArray`, and `filterPositiveInt` are great examples of utility functions. they do something general and useful, and they couldn't be considered 'business logic' that should be in some specific part of the codebase. 

re: `components/utils` -> `components/shared`, agreed. that directory only contains `AutoCompleteSettings.js`, which doesn't really contain _utility_ functionality. `shared` is more on the mark

re: `util` -> `lib`, if anything i would keep `util` and add `lib`. that's a good idea, though. if we're looking for somewhere to put javascript functionality that wouldn't be considered an action/component/container/utility/etc, i think `lib` is as good a name as any

**Comment by gsidebo** [comment]
i also think there could be good reasons to have a `util` file or directory inside an existing directory. take the `actionCreatorGenerator` function you wrote recently for the email front-end. to me, it makes perfect sense to put that in static/js/actions/util.js

**Comment by aliceriot** [comment]
Hm that's an interesting idea. I think that function is really a utility function, whereas stuff like the validation code is really more of a library, it has functionality on it's own and we're using it in a couple places in the app.

So yeah, that makes sense to me. Maybe the deciding factor for utility functions should be that a function belongs in a specific `util` file if it won't be imported outside of that directory, so since `actionCreatorGenerator` won't be used outside of `static/js/actions` it would live in that directory's util file.

Then utility functions used app-wide would still live in `util/util.js`, and libraries that stand on their own more (like the API, the validation code, the profile editing code) would move into `lib`.
